### ISSUE TITLE
New clusters not picked up (Cert Manager http-01 challenge services)

### ISSUE THREAD

**Issue by joonathan** [heptio/contour#304]
When using Contour together with Cert Manager I have faced the following `unknown cluster` issue that required restarting Contour pods.

I verified that the Service was created and was also accessible from within the Cluster.

```
contour-6c6558b87b-gcfhf envoy [2018-03-27 13:25:57.538][15][debug][http] source/common/http/conn_manager_impl.cc:186] [C2014] new stream
contour-6c6558b87b-gcfhf envoy [2018-03-27 13:25:57.538][15][debug][http] source/common/http/conn_manager_impl.cc:790] [C2014][S17148777906128509625] request end stream
contour-6c6558b87b-gcfhf envoy [2018-03-27 13:25:57.538][15][debug][http] source/common/http/conn_manager_impl.cc:455] [C2014][S17148777906128509625] request headers complete (end_stream=true):
contour-6c6558b87b-gcfhf envoy [2018-03-27 13:25:57.538][15][debug][http] source/common/http/conn_manager_impl.cc:460] [C2014][S17148777906128509625]   ':authority':'<REMOVED>'
contour-6c6558b87b-gcfhf envoy [2018-03-27 13:25:57.538][15][debug][http] source/common/http/conn_manager_impl.cc:460] [C2014][S17148777906128509625]   'user-agent':'Go-http-client/1.1'
contour-6c6558b87b-gcfhf envoy [2018-03-27 13:25:57.538][15][debug][http] source/common/http/conn_manager_impl.cc:460] [C2014][S17148777906128509625]   'accept-encoding':'gzip'
contour-6c6558b87b-gcfhf envoy [2018-03-27 13:25:57.538][15][debug][http] source/common/http/conn_manager_impl.cc:460] [C2014][S17148777906128509625]   ':path':'/.well-known/acme-challenge/<REMOVED>'
contour-6c6558b87b-gcfhf envoy [2018-03-27 13:25:57.538][15][debug][http] source/common/http/conn_manager_impl.cc:460] [C2014][S17148777906128509625]   ':method':'GET'
contour-6c6558b87b-gcfhf envoy [2018-03-27 13:25:57.538][15][debug][router] source/common/router/router.cc:236] [C2014][S17148777906128509625] unknown cluster 'prestaging/cm-darwin-prestaging-contour-tls-bhcga/8089'
contour-6c6558b87b-gcfhf envoy [2018-03-27 13:25:57.538][15][debug][http] source/common/http/conn_manager_impl.cc:963] [C2014][S17148777906128509625] encoding headers via codec (end_stream=true):
contour-6c6558b87b-gcfhf envoy [2018-03-27 13:25:57.538][15][debug][http] source/common/http/conn_manager_impl.cc:968] [C2014][S17148777906128509625]   ':status':'503'
contour-6c6558b87b-gcfhf envoy [2018-03-27 13:25:57.538][15][debug][http] source/common/http/conn_manager_impl.cc:968] [C2014][S17148777906128509625]   'date':'Tue, 27 Mar 2018 13:25:57 GMT'
contour-6c6558b87b-gcfhf envoy [2018-03-27 13:25:57.538][15][debug][http] source/common/http/conn_manager_impl.cc:968] [C2014][S17148777906128509625]   'server':'envoy'
```

After booting Contour containers the cluster was picked up and Cert Manager was able to present http-01 challenge and issue certificates.

```
Client Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.6", GitCommit:"6260bb08c46c31eea6cb538b34a9ceb3e406689c", GitTreeState:"clean", BuildDate:"2017-12-21T06:34:11Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"9+", GitVersion:"v1.9.4-gke.1", GitCommit:"10e47a740d0036a4964280bd663c8500da58e3aa", GitTreeState:"clean", BuildDate:"2018-03-13T18:00:36Z", GoVersion:"go1.9.3b4", Compiler:"gc", Platform:"linux/amd64"}
```

running `gcr.io/heptio-images/contour:master` and `docker.io/envoyproxy/envoy-alpine:v1.6.0`

**Comment by davecheney** [comment]
Thanks for raising this issue. I suspect it is related to the pattern of "envoy doesn't see my endpoints" issues which have cropped up over the last week. I'm investigating now on several of those issues but as yet don't have a resolution of workaround.

Would you be able to do some bisecting for me and try the `v0.4.0` and `v0.3.1` tags to confirm if the issue is present for you in earlier releases?

**Comment by joonathan** [comment]
I’ll try to experiment with previous versions in the dev cluster in the morning. Thanks for looking into it.

**Comment by davecheney** [comment]
Thanks to Alexander Lukyanchenko (@Lookyan) we have increased the general gRPC limits on both the Envoy client and Contour server well above anything that should be an issue for the immediate future.

The symptoms of hitting gRPC limits vary, but are basically "envoy doesn't see changes in the API server until I restart <things>".
The underlying cause is likely to be that you have a large (more than 100, possibly 200, the exact limit is not precisely known) number of Service objects in your cluster -- these don't have to be associated with an Ingress.
Currently Contour creates a CDS Cluster record for any Service object it learns about through the API, see #298.
Each CDS record will cause Envoy to open a new EDS stream, one per Cluster, which can blow through the default limits that Envoy, as the gRPC client, and Contour, as the gRPC server, have set.

One of the easiest ways to detect if this issue is occuring in your cluster is too look for lines about "cluster warming"
```
[2018-04-03 03:34:16.920][1][info][upstream] source/common/upstream/cluster_manager_impl.cc:388] add/update cluster test2/reverent-noether/80 starting warming
[2018-04-03 03:34:16.922][1][info][upstream] source/common/upstream/cluster_manager_impl.cc:388] add/update cluster test2/serene-bohr/80 starting warming
[2018-04-03 03:34:16.924][1][info][upstream] source/common/upstream/cluster_manager_impl.cc:388] add/update cluster test2/sleepy-hugle/80 starting warming
```
Without a matching "warming complete" message.

We believe we have addressed this issue, #291, and the fixes are available now to test.

These changes are in master now, and available in the gcr.io/heptio-images/contour:master image for you to try.

This has been backported the release-0.4 branch and are available in a short lived image gcr.io/heptio-images/contour:release-0.4 (it's not going to be deleted, but don't expect it to continue to be updated beyond the 0.4.1 release).
